### PR TITLE
[train+tune] Local directory refactor (2/n): Separate driver artifacts and trial working directories

### LINKED ISSUE #38522: [Train] Ray Train sync all previous trial directories from local to remote storage
### Description

I was using NFS as storage path first, and generated a bunch of directories. Then I change the storage path to a S3 bucket, and run it once (`TorchTrainer_a006b_00000_0_2023-08-16_14-26-55/`). My hypothesis is that only the current trial dir will appear in the S3 storage path, but it seems all the previous trial directories are all synced to S3.

It turns out that we have a local cache dir, which maintains a copy of all previous runs locally. Train internal will automatically sync all the contents under the same experiment name (`ray-deepspeed-demo` here) to S3.

```
(base) ray@ip-10-0-41-0:~$ aws s3 ls s3://anyscale-staging-data-cld-kvedzwag2qa8i5bjxuevf5i7/org_7c1Kalm9WcX2bNIjW53GUT/cld_kvedZWag2qA8i5BjxUevf5i7/artifact_storage/yx_ray_results/ray-deepspeed-demo/
                           PRE TorchTrainer_21cdf_00000_0_2023-08-16_14-09-05/
                           PRE TorchTrainer_46979_00000_0_2023-08-16_14-17-16/
                           PRE TorchTrainer_842d8_00000_0_2023-08-16_14-11-50/
                           PRE TorchTrainer_a006b_00000_0_2023-08-16_14-26-55/
                           PRE TorchTrainer_a0195_00000_0_2023-08-16_14-12-37/
                           PRE TorchTrainer_ce826_00000_0_2023-08-16_14-21-04/
2023-08-16 14:28:58          0 
2023-08-16 14:28:58          0 .validate_storage_marker
2023-08-16 14:29:03       7912 basic-variant-state-2023-08-16_14-09-05.json
2023-08-16 14:29:03       7928 basic-variant-state-2023-08-16_14-11-50.json
2023-08-16 14:29:03       7912 basic-variant-state-2023-08-16_14-12-37.json
2023-08-16 14:29:04       7912 basic-variant-state-2023-08-16_14-17-16.json
2023-08-16 14:29:04       7514 basic-variant-state-2023-08-16_14-21-04.json
2023-08-16 14:29:04       8638 basic-variant-state-2023-08-16_14-26-55.json
2023-08-16 14:29:04      16590 experiment_state-2023-08-16_14-09-05.json
2023-08-16 14:29:04       8565 experiment_state-2023-08-16_14-11-50.json
2023-08-16 14:29:05       8586 experiment_state-2023-08-16_14-12-37.json
2023-08-16 14:29:05      21100 experiment_state-2023-08-16_14-17-16.json
2023-08-16 14:29:05       8222 experiment_state-2023-08-16_14-21-04.json
2023-08-16 14:29:05      23299 experiment_state-2023-08-16_14-26-55.json
2023-08-16 14:29:05       5148 trainer.pkl
2023-08-16 14:29:06       2382 tuner.pkl
```

### Use case

_No response_

### LINKED ISSUE #40009: [Tune/Air] Setting a custom storage_path in air.RunConfig creates duplicates in default and specified storage directory
### What happened + What you expected to happen

**1. the bug:**
I'm running a RLlib reinforcement learning algorithm using ray.tune with a user-specified `storage_path` and `name` in the air.RunConfig, as shown in the minimal example. 
Instead of having the results in `storage_path/name`, another results directory gets created in `~/ray_results/name`. The latter directory also gets updated with TensorBoard results during the training run. Only at the end of training, the TensorBoard file (as well as the checkpoint directory) appear in the correct directory.

**2. expected behavior:**
the results are only supposed to be created in the specified storage path.

### Versions / Dependencies

Ray: 2.7.0
Python: 3.9.18
OS: Ubuntu 22.04.02

### Reproduction script

```
from ray import air, tune
from ray.rllib.algorithms.ppo import PPOConfig


algo_config = PPOConfig().environment("CartPole-v1").framework("torch")

tuner = tune.Tuner(
    "PPO",
    param_space=algo_config.to_dict(),
    run_config=air.RunConfig(
        name="run_01",
        storage_path="~/ray_results/CartPole-v1",
        stop={"episode_reward_mean": 475},

    ),
)

tuner.fit()
```

### Issue Severity

Low: It annoys or frustrates me.

### LINKED ISSUE #40634: [train] Logs from frameworks (`lightning_logs`, `wandb`, transformers `output_dir`) in the working directory can be synced unintentionally
Many frameeworks set default logging directories to the working directory.

Train/Tune changes the working directory to the trial directory, and the contents of this directory can get synced to cloud unintentionally. This can cause double uploading of checkpoints (once for the Train checkpoint, and once as an artifact in the directory).

The uploading happens from either:
1. Driver syncing if the trial happens to live on the head node. This can be fixed by converting the sync exclude-list into an explicit include-list instead.
2. Trial artifact syncing enabled by `SyncConfig(sync_artifacts=True)`. We should either recommend to configure the logging directory of these frameworks to an external directory in the docs, or add a configurable artifact exclude-list.

### PR DESCRIPTION
**Pull request by justinvyu**
## Change summary

* Behavior changes:
    * **Rebrands the `local_dir` concept as a _driver staging directory_ and moves its default location from `~/ray_results` to a subfolder in the ray temp directory (`/tmp/ray/session_*/artifacts`).**
        * Delegates the customization of this folder to `ray.init(_temp_dir=...)`.
        * Increases the visibility of this staging directory so that users can find and delete it if they want to. Ray already logs things to this directory, so it's a natural directory for users to interact with.
    * **Provides trial actors (Tune) and distributed training workers (Train) an empty working directory that will only be populated by worker artifacts.**
        * This working directory is the `storage.trial_working_directory` and will only be synced if `sync_artifacts=True`.
        * This is the same behavior as before, with the added benefit that driver syncing no longer unintentionally uploads worker artifacts.
    * **Simplifies the default `storage_path` resolution.**
        * `storage_path` is now resolved immediately on `RunConfig` initialization, rather than resolving downstream during `tune.run`. It gets set to the Ray storage URI if that's setup -- otherwise, it defaults to `~/ray_results`.
        * Prior to this PR, `~/ray_results` was already the default location of `storage_path`. The difference now is that `~/ray_results` no longer be populated if the `storage_path` is set to something else.
    * **Syncing from the staging directory to the storage path is now the only codepath.**
        * There is no more `syncer=None` special case when `local_dir == storage_path`.
        * This means that for single node runs, there is some unnecessary file-copying, but this is an ok sacrifice for simplifying the codepaths.
    * **Experiment checkpoint saving and uploading now happens synchronously.**
        * Previously, this happened in 2 steps: (1) the files would get saved to `local_dir`, (2) they get uploaded to `storage_path` in a background task (this is "driver syncing").
        * The frequency of the 2nd step gets gated by `SyncConfig(sync_timeout)`, which is 5 minutes by default. Now that the local directory only contains driver artifacts, the upload is not so expensive, making it okay to upload synchronously right after saving.
        * The benefit of this is better consistency of the experiment state in `storage_path`.
        * I also removed the forced syncing behavior of `num_to_keep` now that uploading driver files is no longer gated by the 5 minute `sync_period` default. That workaround was intended to increase driver checkpointing frequency, which is also achieved by change.
        * See [1](https://github.com/ray-project/ray/pull/43403#discussion_r1512036893) and [2](https://github.com/ray-project/ray/pull/43403#discussion_r1501214437) for context on the motivation for this change.
* Some undesirable workarounds made in this PR:
    * Adding a unique timestamp to the path of the driver staging directory so that different experiments with the same `RunConfig(name)` don't have conflicting staging directories.
        * This was the issue that needed to be solved: https://github.com/ray-project/ray/pull/43369#discussion_r1508190626
        * Now, `StorageContext` gets a property of the current timestamp upon creation (which happens on the driver once).
        * The storage context gets propagated to the remote trial actors and shares the same timestamp.
        * On restoration, a new storage context gets created on the driver, which is now different from the timestamp of the storage context on the restored trial.
        * We workaround this by [updating the timestamp of the driver's storage context with the restored timestamp.](https://github.com/ray-project/ray/pull/43403#discussion_r1512290914)
        * A restored run will use the same staging directory as the original run. A new run will always use a new timestamp.
    * The `/tmp/ray/session_*` folder is only available when ray init has been called. I included the ray start fixture in many unit tests to get around this. `mock_storage_context` patches out this directory to a tempdir.
* Remaining TODOs for follow-ups:
    * A hanging/failing driver file upload will currently block/fail the tune control loop, even though all trials may be running fine. This is a regression compared to the previous behavior of just logging a warning and retrying the upload. This will be fixed in https://github.com/ray-project/ray/pull/43689.
    * [Possibly re-introduce a way to modify the Ray Train staging directory without needing to move the entire Ray temp dir.](https://github.com/ray-project/ray/pull/43403#discussion_r1508258561)
    * [Clean up `Trial` and `Experiment` path properties once and for all.](https://github.com/ray-project/ray/pull/43403#discussion_r1508294469)
    * Removing usage/mentions of `RAY_AIR_LOCAL_CACHE_DIR` everywhere.
    * Docs updates:
        * Tune relative path FAQ section. `ray.train.get_context().get_local_dir` doesn't really make sense anymore.
        * Update the persistent storage user guide.

## Context / motivation for this change

What are the main user problems to solve?
* Problem 1: ~/ray_results gets populated and it’s unclear how this thing is configured and how it can be disabled.
    * https://github.com/ray-project/ray/issues/42630
    * https://github.com/ray-project/ray/issues/40009
    * https://ray-distributed.slack.com/archives/CSX7HVB5L/p1698324789112559
    * https://ray-distributed.slack.com/archives/CNECXMW22/p1694712269714009
    * **Problem 1 gets solved by hiding away the `~/ray_results` to the ray session dir that already gets populated and is easily accessible / configurable.**
* Problem 2: Implicit uploading of CWD logs from driver syncing causes slow syncing and unintended files getting uploaded.
    * https://github.com/ray-project/ray/issues/40634
    * **Problem 2 gets solved by separating the driver artifact folders and the trial working directories in different folders.**
* Problem 3: If you keep using the same `RunConfig(name)` and run multiple experiments, each consecutive experiment will upload ALL “trial directories” so far from the local `~/ray_results`.
    * Ex: exp 1 has 1 trial folder, exp 2 has 2 (the current one plus the previous one), etc.
    * https://github.com/ray-project/ray/issues/38522
    * **Problem 3 gets solved by using the latest ray session dir, so that every run has a blank slate rather than accumulating in a single `~/ray_results/<name>` folder.**

## Related issue number

Closes https://github.com/ray-project/ray/issues/40634
Closes https://github.com/ray-project/ray/issues/40009
Closes https://github.com/ray-project/ray/issues/38522

https://github.com/ray-project/ray/issues/42630

## Checks

- [ ] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
    - [ ] I've added any new APIs to the API Reference. For example, if I added a 
           method in Tune, I've added it in `doc/source/tune/api/` under the 
           corresponding `.rst` file.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: python/ray/train/_internal/storage.py [hunk 1]
```diff
@@ -429,46 +421,31 @@ def __init__(
...
-
         # If no remote path is set, try to get Ray Storage URI
         ray_storage_uri: Optional[str] = _get_storage_uri()
-        if ray_storage_uri and storage_path is None:
+        if ray_storage_uri is not None:
             logger.info(
                 "Using configured Ray Storage URI as the `storage_path`: "
                 f"{ray_storage_uri}"
             )
+            storage_path = ray_storage_uri
```
Comment (justinvyu) [inline python/ray/train/_internal/storage.py:431]: Should we continue supporting Ray Stroage? Previously, we fall back to ray storage if `storage_path` is not set.

Now, I'm trying to simplify `storage_path` so that it's always set in the `RunConfig` directly -- defaulting to `~/ray_results`. 

This means that it'll never be `None` / auto-resolved anymore. So, the behavior here has changed to: if ray storage is setup, then override the `storage_path`. Thoughts on this change? Or alternatively we can drop this behavior entirely?

File: python/ray/train/_internal/storage.py [hunk 2]
```diff
@@ -614,37 +590,85 @@ def experiment_fs_path(self) -> str:
         return Path(self.storage_fs_path, self.experiment_dir_name).as_posix()
 
     @property
-    def experiment_local_path(self) -> str:
-        """The local filesystem path to the experiment directory.
+    def experiment_local_staging_path(self) -> str:
```
Comment (justinvyu) [inline python/ray/train/_internal/storage.py:593]: Thoughts on the var naming / directory structure?

File: python/ray/train/tests/test_new_persistence.py [hunk 3]
```diff
@@ -416,13 +407,12 @@ def _assert_storage_contents(
 
 
 @pytest.mark.parametrize("trainable", [train_fn, ClassTrainable])
-@pytest.mark.parametrize("storage_path_type", [None, "nfs", "cloud", "custom_fs"])
```
Comment (justinvyu) [inline python/ray/train/tests/test_new_persistence.py:419]: We don't need to test `None` anymore since it's the same as the "nfs" test.

File: python/ray/tune/experiment/trial.py [hunk 4]
```diff
@@ -536,7 +536,7 @@ def remote_experiment_path(self) -> str:
 
     @property
     def local_experiment_path(self) -> str:
-        return self.storage.experiment_local_path
+        return self.storage.experiment_local_staging_path
```
Comment (justinvyu) [inline python/ray/tune/experiment/trial.py:539]: Should we delete all of these `Trial` properties and just always access paths through the `storage`? Neither are really stable APIs.

Comment (justinvyu) [reply python/ray/tune/experiment/trial.py:539]: Same thing for the `Experiment` class

File: python/ray/tune/execution/tune_controller.py [hunk 5]
```diff
@@ -338,43 +336,39 @@ def checkpoint_exists(cls, directory: str) -> bool:
...
-        os.replace(
-            tmp_file_name,
-            Path(experiment_dir, self.experiment_state_file_name).as_posix(),
+        experiment_local_staging_path = self._storage.experiment_local_staging_path
+        self._search_alg.save_to_dir(
+            experiment_local_staging_path, session_str=self._session_str
+        )
+        self._callbacks.save_to_dir(
+            experiment_local_staging_path, session_str=self._session_str
         )
```
Comment (justinvyu) [inline python/ray/tune/execution/tune_controller.py:371]: This change uploads just the TuneController to state to storage directly, rather than dumping it to a file and waiting for the driver sync to upload it.

Since searcher state saving is really complicated, I still dump those files and upload them implicitly through driver sync uploading.

The reason I made this change is to fix some broken tests that weren't uploading the `TuneController` state often enough. The local "staging" version of the file was up to date, but it wasn't being uploaded in time to restore from the right state after a crash on the driver (ex: a callback raised an error).

File: python/ray/tune/tests/test_tuner_restore.py [hunk 6]
```diff
@@ -338,12 +338,6 @@ def test_tuner_resume_unfinished(ray_start_2_cpus, tmpdir, monkeypatch):
     """Resuming unfinished trials should pick up existing state"""
     monkeypatch.setenv("TUNE_GLOBAL_CHECKPOINT_S", "0")
 
-    # TODO(justinvyu): Setting storage_path to this tempdir causes this test to fail.
-    # This is because the error raised by the driver callback doesn't let the
-    # experiment sync happen (from ~/ray_results -> tmpdir). This would also
-    # be the case for real cloud syncing.
-    monkeypatch.setenv("RAY_AIR_LOCAL_CACHE_DIR", str(tmpdir))
-
```
Comment (justinvyu) [inline python/ray/tune/tests/test_tuner_restore.py:346]: Example test that works now after the direct exp state upload.

File: python/ray/tune/tests/test_tuner_restore.py [hunk 7]
```diff
@@ -397,17 +391,14 @@ def test_tuner_resume_unfinished(ray_start_2_cpus, tmpdir, monkeypatch):
 
     results = tuner.fit()
     assert len(results) == 4
-    assert len(results.errors) == 1
+    # assert len(results.errors) == 1
```
Comment (woshiyyya) [inline python/ray/tune/tests/test_tuner_restore.py:394]: why we ignored the error here?

File: python/ray/tune/experiment/trial.py (hunk 4 above)
Comment (woshiyyya) [reply python/ray/tune/experiment/trial.py:539]: Yeah, the naming of these properties are kind of confusing. We can deprecate these properties and take `StorageContext` as ground truth.

File: python/ray/train/_internal/storage.py (hunk 2 above)
Comment (woshiyyya) [reply python/ray/train/_internal/storage.py:593]: what about putting the driver concept into the name? e.g. `driver_experiment_staging_dir`?

File: python/ray/train/_internal/storage.py [hunk 8]
```diff
@@ -614,37 +590,85 @@ def experiment_fs_path(self) -> str:
...
+
+    @property
+    def trial_working_directory(self) -> str:
+        """The local filesystem path to trial working directory.
+
+        This path is of the form:
+        `/tmp/ray/session_<session_id>/artifacts/<experiment_dir_name>/working_dirs/<trial_dir_name>`
+
+        Ray Train/Tune moves the remote actor's working directory to this path
+        by default, unless disabled by `RAY_CHDIR_TO_TRIAL_DIR` environment variable.
```
Comment (woshiyyya) [inline python/ray/train/_internal/storage.py:656]: Mention the result trial dir abs path?
`{RAY_CHDIR_TO_TRIAL_DIR}/working_dirs/<trial_dir_name>` or `{RAY_CHDIR_TO_TRIAL_DIR}/<trial_dir_name>`?

File: python/ray/tune/experiment/trial.py (hunk 4 above)
Comment (justinvyu) [reply python/ray/tune/experiment/trial.py:541]: Ok, I'll do it in a followup PR (doesn't have to make branch cut).

File: python/ray/train/_internal/storage.py (hunk 2 above)
Comment (justinvyu) [reply python/ray/train/_internal/storage.py:595]: Yeah, that might be clearer. I'll make the change and label this property as experimental -- users should not access it / rely on it.

File: python/ray/train/_internal/storage.py (hunk 1 above)
Comment (justinvyu) [reply python/ray/train/_internal/storage.py:431]: Ah, I thought of a solution -- we can do this check directly in the `RunConfig` init.

File: python/ray/train/_internal/session.py [hunk 9]
```diff
@@ -222,15 +222,14 @@ def reset(
         self.training_started = False
         self._first_report = True
 
-        # Change the working directory to the local trial directory.
-        # -> All workers on the same node share a working directory.
-        os.makedirs(storage.trial_local_path, exist_ok=True)
```
Comment (woshiyyya) [inline python/ray/train/_internal/session.py:227]: I am actually a bit worried that we changed the working dir into the ray global session dir in the performance perspective:

For example, users may save large checkpoints into `/tmp` folder, and use up the disk storage. Intuitively we should only save the small metadata and logs in this Ray session dir.

A mitigation is to change ray global session dir onto another disk (e.g. mounted SSD with larger storage and higher bandwidth), all other logs (ray core, data, ...) will be moved together in consequence. We probably don't want our ray train featuer affects other libraries' behavior.

File: python/ray/train/_internal/storage.py [hunk 10]
```diff
@@ -614,37 +592,92 @@ def experiment_fs_path(self) -> str:
...
+    def trial_working_directory(self) -> str:
+        """The local filesystem path to trial working directory.
 
-        Raises a ValueError if `trial_dir_name` is not set beforehand.
+        This path is of the form:
+        `/tmp/ray/session_<session_id>/artifacts/<ray-train-job-timestamp>/
+        <experiment_dir_name>/working_dirs/<trial_dir_name>`
+
+        Ray Train/Tune moves the remote actor's working directory to this path
+        by default, unless disabled by `RAY_CHDIR_TO_TRIAL_DIR` environment variable.
```
Comment (woshiyyya) [inline python/ray/train/_internal/storage.py:664]: What would happen if `RAY_CHDIR_TO_TRIAL_DIR=0`? I saw we are always using  `_get_ray_train_session_dir()`.

File: python/ray/tune/execution/tune_controller.py [hunk 11]
```diff
@@ -519,6 +428,84 @@ def resume(self, resume_config: ResumeConfig):
...
+
+            # ATTN: `trial.set_storage` is used intentionally, since it
+            # also updates the absolute paths and filesystem of tracked checkpoints.
+            trial.set_storage(new_storage)
+
+            # Avoid creating logdir in client mode for returned trial results,
+            # since the dir might not be creatable locally.
+            # TODO(ekl) this is kind of a hack.
+            if not ray.util.client.ray.is_connected():
+                trial.init_local_path()  # Create logdir if it does not exist
```
Comment (woshiyyya) [inline python/ray/tune/execution/tune_controller.py:454]: The naming is kind of convoluted now. In the future, we should have a unified naming, e.g. all use `staging` instead of `local`.

File: python/ray/train/_internal/storage.py (hunk 1 above)
Comment (justinvyu) [reply python/ray/train/_internal/storage.py:431]: `ray_storage_uri` is set through `ray.init(..., storage="s3://...")`, so users will have already called ray init.

File: python/ray/train/tests/test_new_persistence.py [hunk 12]
```diff
@@ -575,15 +558,22 @@ def test_trainer(
         ├── artifact-rank=1-iter=1.txt
         └── ...
     """
-    LOCAL_CACHE_DIR = tmp_path / "ray_results"
-    monkeypatch.setenv("RAY_AIR_LOCAL_CACHE_DIR", str(LOCAL_CACHE_DIR))
```
[the remark this answers is not in the document]
Comment (justinvyu) [reply python/ray/train/tests/test_new_persistence.py:579]: I patched the local dir before so that nothing got written to `~/ray_results`, but instead a temp directory that is unique per test. I didn't want the staging directories of multiple tests to be mixed.

File: python/ray/train/_internal/session.py (hunk 9 above)
Comment (justinvyu) [reply python/ray/train/_internal/session.py:227]: `/tmp/ray` is the place where large ray objects get spilled, and I am also a bit worried that random things written by to the working directory will cause out of disk on the `/tmp` drive.

For now, it's possible to work around this by setting a different output directory for saving checkpoints, so let's just keep it as is. The other driver files are very small, so it's fine for the general case.

In the future, if there are user issues related to this, we can consider adding an environment variable to allow users to set the ray train staging path without needing to move all of the ray logs.

File: python/ray/train/_internal/storage.py (hunk 10 above)
Comment (justinvyu) [reply python/ray/train/_internal/storage.py:664]: The remote actors will just not be moved to this directory. However, this is still the directory that would be synced to storage if `sync_artifacts=True`. So, users would need to write files to `storage.trial_working_directory` if they wanted to use that feature.

I need to upload this faq section: https://docs.ray.io/en/latest/tune/faq.html#how-do-i-access-relative-filepaths-in-my-tune-training-function

File: python/ray/tune/tests/test_tuner_restore.py (hunk 7 above)
Comment (justinvyu) [reply python/ray/tune/tests/test_tuner_restore.py]: Fixed.

File: python/ray/air/config.py [hunk 13]
```diff
@@ -634,8 +634,13 @@ class RunConfig:
 
     def __post_init__(self):
         from ray.train import SyncConfig
+        from ray.train.constants import DEFAULT_STORAGE_PATH
         from ray.tune.experimental.output import AirVerbosity, get_air_verbosity
 
+        if self.storage_path is None:
+
+            self.storage_path = DEFAULT_STORAGE_PATH
```
[the remark this answers is not in the document]
Comment (justinvyu) [reply python/ray/air/config.py:642]: I will update the storage user guide in a followup docs PR.

File: python/ray/train/_internal/session.py (hunk 9 above)
Comment (woshiyyya) [reply python/ray/train/_internal/session.py:227]: Yes it'd be better to allow users to specify a custom ray train staging path. Let's create an issue/add a todo?

**Review by woshiyyya — COMMENTED** [review]
Nice to see the tests become green!

As this is a large PR changed a bunch of behaviors and tests. Can you summarize the following things in the PR descriptions?
- Behavior changes
  - changes in the way of driver syncing (Directly write to storage path)
  - how to separate trial and driver artifact directory
  - changes in the default storage path (`ray_storage_uri`)
  - ...
- The compromise we've done
  - The bypassed unit tests
  -  sync_artifacts
  - ...
- The TODOs  
  - doc and faqs to update
  -  

Also, before we merge the PR, It'd also be good to run a few release tests to ensure it also works under multi-node setting.

File: python/ray/train/_internal/storage.py [hunk 14]
```diff
@@ -451,6 +452,10 @@ def __init__(
         self._create_validation_file()
         self._check_validation_file()
 
+        # Timestamp is used to create a unique session directory for the current
+        # training job. This is used to avoid conflicts when multiple training jobs
+        # run with the same name in the same cluster.
+        # This is set ONCE at the creation of the storage context, on the driver.
```
Comment (woshiyyya) [inline python/ray/train/_internal/storage.py:458]: Can you help me remember how did we resolved the consistency issue with the timestamp. Did we resolve it by writing files into driver directory and using background syncing.

**Review by woshiyyya — APPROVED** [review]

File: python/ray/tune/execution/tune_controller.py (hunk 5 above)
Comment (justinvyu) [reply python/ray/tune/execution/tune_controller.py]: This solution has been replaced with dumping to the directory + uploading immediately after.

This is because searcher state is too hard to refactor -- the implementation of save for many searchers is tied to writing files.

See https://github.com/ray-project/ray/pull/43403/files#diff-56de03a86c3b7df66c61730aba6813a52bf26cbb424357b9bd35b7294b18294bR185-R189

File: python/ray/tune/execution/tune_controller.py [hunk 15]
```diff
@@ -519,6 +429,84 @@ def resume(self, resume_config: ResumeConfig):
...
+
+            trials.append(trial)
+
+        # NOTE: The restored run should reuse the same driver staging directory.
+        self._storage._timestamp = trials[0].storage._timestamp
```
Comment (justinvyu) [inline python/ray/tune/execution/tune_controller.py:460]: Workaround to make sure that a restored run uses the same timestamped staging dir.

File: python/ray/train/_internal/storage.py (hunk 14 above)
Comment (justinvyu) [reply python/ray/train/_internal/storage.py:458]: See the first bullet point in the PR description under "Some undesirable workarounds made in this PR."

**Comment by justinvyu** [comment]
`train_multinode_persistence` release test passed here: https://buildkite.com/ray-project/release/builds/10131#018e0d55-eb73-42aa-abe9-e3ba603a61f5

I removed the `tune_cloud_durable_upload` test and need to add back a tune multi-node persistence test as a follow-up.
